Lib/slapdtest/_slapdtest.py

```diff
@@ -483,34 +482,57 @@ def _stopped(self):
             self._log.info('slapd[%d] terminated', self._proc.pid)
             self._proc = None
 
-    def _cli_auth_args(self):
-        if self.cli_sasl_external:
-            authc_args = [
-                '-Y', 'EXTERNAL',
-            ]
-            if not self._log.isEnabledFor(logging.DEBUG):
-                authc_args.append('-Q')
-        else:
-            authc_args = [
-                '-x',
-                '-D', self.root_dn,
-                '-w', self.root_pw,
-            ]
-        return authc_args
+    def _cli_slap(self, cmd=None, extra_args=None, stdin_data=None):  # pragma: no cover
+        """Entry point for slap* commands directly modifying DB files."""
+        if self._proc is not None:
+            self._log.warning('Directly modifying DB files while slapd is running')
+
+        args = [self.PATH_SLAPD]
+
+        if cmd:
+            args.append('-T'+cmd)
+
+        args += ['-F', self._slapd_conf] + (extra_args or [])
+
+        return self._cli_exec(args, stdin_data=stdin_data)
+
+    def _cli_ldap(self, ldapcommand, extra_args=None, ldap_uri=None,
+                  bind_dn=None, bind_pw=None,
+                  stdin_data=None):  # pragma: no cover
+        """
+        Entry point for ldap* commands, interacting with running slapd
+        """
+        if self._proc is None:
+            raise RuntimeError('Must start daemon before LDAP access is possible')
 
-    # no cover to avoid spurious coverage changes
-    def _cli_popen(self, ldapcommand, extra_args=None, ldap_uri=None,
-                   stdin_data=None):  # pragma: no cover
         if ldap_uri is None:
             ldap_uri = self.default_ldap_uri
 
-        if ldapcommand.split("/")[-1].startswith("ldap"):
-            args = [ldapcommand, '-H', ldap_uri] + self._cli_auth_args()
+        args = [ldapcommand, '-H', ldap_uri or self.default_ldap_uri]
+
+        if ldap_uri.startswith('ldapi://'):
+            args += ['-Y', 'EXTERNAL']
+
+            if not self._log.isEnabledFor(logging.DEBUG):
+                args.append('-Q')
+
+            if bind_dn or bind_pw:
+                raise RuntimeError('-Y EXTERNAL ignores -D and -w')
```

Comment:
How about treating those being passed in as a signal to use a simple bind? Also the `self.cli_sasl_external` check is gone?

Reply:
> How about treating those being passed in as a signal to use a simple bind?

Maybe showing my ignorance here.  Is a simple bind meaningful with `ldapi://`?

So the condition for choosing `-Y EXTERNAL` would be `ldap_uri.startswith('ldapi://') and bind_dn is None`?

> Also the `self.cli_sasl_external` check is gone?

Yes.  I don't think this was the correct condition.  Knowing that `ldapi://` is supported doesn't imply that it will be used for a particular operation.

Reply:
The SASL `EXTERNAL` mechanism is independent of the `ldapi://` transport. `cli_sasl_external` tells you whether that mechanism is available/should be used, `ldapi://` tells you how to contact the server (over a UNIX socket vs. as opposed to a TCP socket), the LDAP protocol still behaves the same over both.

Hope that clears things up a bit, let me know if not.

Reply:
For example if we set up a TLS layer and used a client certificate while doing so, some servers could also accept the `EXTERNAL` mechanism, using the client certificate to generate the identity the session should bind as (OpenLDAP does this).

BTW not asking that you take it into account in this pull request, just giving some background.